clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -4490,6 +4490,50 @@ def CIR_TryOp : CIR_Op<"try",[
   let hasLLVMLowering = false;
 }
 
+//===----------------------------------------------------------------------===//
+// CatchParamOp
+//===----------------------------------------------------------------------===//
+
+def CIR_CatchParamKind  : CIR_I32EnumAttr<
+  "CatchParamKind", "Designate limits for begin/end of catch param handling", [
+    I32EnumAttrCase<"Begin", 0, "begin">,
+    I32EnumAttrCase<"End",  1, "end">
+]>;
+
+def CIR_CatchParamOp : CIR_Op<"catch_param"> {
+  let summary = "Represents catch clause formal parameter";
+  let description = [{
+    The `cir.catch_param` can operate in two modes: within catch regions of
+    `cir.try` or anywhere else with the `begin` or `end` markers. The `begin`
+    version requires an exception pointer of `cir.ptr<!void>`.
```

Comment:
It might be worth expanding on how the two modes are used (one that assumes structured control and the other one for lower level lowering). An alternative design would be split this into two operations (let's see if others have any strong opinions here).

Reply:
Happy to approve this PR once this comment is addressed

Reply:
Am I understanding correctly that the difference between the modes is that the begin/end mode is the result of CFG flattening of the other mode?

I'm not clear what the purpose of the begin/end markers is. Is it related to Windows funclet generation? Or does it just delineate a logical region across arbitrary control flow?

I think I'd prefer to have two different operations for the modes.